src/identity.js

```diff
@@ -366,62 +389,86 @@ export class Identity extends EventEmitter {
      * @fires Identity#sessionInit
      * @fires Identity#statusChange
      * @fires Identity#error
-     * @return {Identity#HasSessionSuccessResponse|Identity#HasSessionFailureResponse}
+     * @return {Promise<Identity#HasSessionSuccessResponse|Identity#HasSessionFailureResponse>}
      */
-    async hasSession(autologin = true) {
-        const loginInProgress = this.cache.get(LOGIN_IN_PROGRESS_KEY) !== null;
-        this.cache.delete(LOGIN_IN_PROGRESS_KEY);
-
-        const postProcess = (sessionData) => {
-            if (sessionData.error) {
-                throw new SDKError('HasSession endpoint returned an error', sessionData.error);
-            }
-            this._maybeSetVarnishCookie(sessionData);
-            this._emitSessionEvent(this._session, sessionData);
-        };
-
-        if (typeof autologin !== 'boolean') {
-            const [type, value] = inspect(autologin);
-            throw new SDKError(`Parameter 'autologin' must be boolean, was: "${type}:${value}"`);
-        }
-        if (this._enableSessionCaching) {
-            // Try to resolve from cache (it has a TTL)
-            const cachedData = this.cache.get(HAS_SESSION_CACHE_KEY);
-            if (cachedData) {
-                postProcess(cachedData);
-                return cachedData;
-            }
-        }
-
-        try {
-            const autoLoginConverted = autologin ? 1 : 0;
-            let data = null;
-            if (!this._itpMode) {
-                data = await this._hasSession.get('rpc/hasSession.js', { autologin: autoLoginConverted });
-            }
-            if (this._itpMode || (isObject(data.error) && data.error.type === 'LoginException')) {
-                data = await this._spid.get('ajax/hasSession.js', { autologin: autoLoginConverted });
-            }
-
-            const shouldShowItpModal = this._itpModalRequired() && !this._itpMode &&
-                isObject(data.error) && data.error.type === 'UserException' &&
-                loginInProgress;
-            if (shouldShowItpModal) {
-                const modal = new ItpModal(this._spid, this.clientId, this.redirectUri, this.env);
-                data = await modal.show();
-            }
-
-            if (this._enableSessionCaching) {
-                const expiresIn = 1000 * (data.expiresIn || 300);
-                this.cache.set(HAS_SESSION_CACHE_KEY, data, expiresIn);
-            }
-            postProcess(data);
-            this._session = data;
-            return data;
-        } catch (err) {
-            this.emit('error', err);
-            throw new SDKError('HasSession failed', err);
+    hasSession(autologin = true) {
+        if (!this._hasSessionInProgress) {
```

Comment:
personally I'd do 
```
if (this._hasSessionInProgress) {
   return _hasSessionInProgress
}

const promiseFn...
```
to cut down on nesting

Reply:
Will do!